### PR TITLE
[WIP] Array op

### PR DESCRIPTION
**Pull request by porglezomp**
This adds arrays, WIP so @a2aaron can implement the actual evaluation.

### CODE REVIEW

**Comment by a2aaron** [comment]
I'm assuming that the operator `[n` pops a value off the stack (which I will call `i`), and then pops n values off the stack (the array portion). It then pushes the value of the number at the `i % n`th index in the array. Is this right?

Additionally,

Is `[0` valid, and if it is, what would it `[0` do?

For example, if we have the stack as `1 2 3 0`, does `[0` just pop nothing, leaving our stack as `1 2 3 0` or does it pop off one value off the top anyway, leaving our stack as `1 2 3`? I personally think it should pop one value off the top, as that would make it easier to implement.

**Comment by porglezomp** [comment]
The stack effect of `[n` is popping `n + 1` items (the `n` for the array, and `1` for the index) and pushing `1` item, so it decreases the size of the stack by `n`. If we pop the index off the top for `[0`, then we need to push something back on. I support always popping the index, and pushing `0` if the size of the array is `0`.

**Comment by porglezomp** [comment]
Make sure to add a test case for the empty array case.

File: src/lib.rs [hunk 1]
```diff
@@ -145,7 +145,26 @@ pub fn eval_beat(cmds: &[Cmd], t: f64) -> Result<f64, ()> {
                     stack.push(a % b);
                 }
             }
-            Arr(size) => unimplemented!(),
+            Arr(size) => {
+                // We need to pop index + size values, so our stack needs to be
+                // atleast size + 1 elements long
+                if size + 1 > stack.len() || size == 0 {
```
Comment (porglezomp) [inline src/lib.rs]: This seems wrong, we should do what we usually do an fail with an `Err(())` if the stack is mismatched.

File: src/lib.rs [hunk 2]
```diff
@@ -145,7 +145,26 @@ pub fn eval_beat(cmds: &[Cmd], t: f64) -> Result<f64, ()> {
...
+            Arr(size) => {
+                // We need to pop index + size values, so our stack needs to be
+                // atleast size + 1 elements long
+                if size + 1 > stack.len() || size == 0 { 
+                    stack.push(0.0)
+                } else {
+                    // We want to split off from the end, so we must subtract
+                    // here. Note that we add one for the index.
+                    let split_index = stack.len() - (size + 1);
+                    let mut vec = stack.split_off(split_index);
```
Comment (porglezomp) [inline src/lib.rs]: We can simplify this calculation by popping the index off first, and then splitting off, so we don't need the `size + 1`, and the later shuffling.

File: src/lib.rs [hunk 3]
```diff
@@ -522,6 +541,30 @@ mod tests {
         }
     }
 
+    test_beat! {
```
Comment (porglezomp) [inline src/lib.rs:697]: I think this behavior is wrong, this should probably be a separate test that asserts that running this is `Err(())`.

File: src/lib.rs [hunk 4]
```diff
@@ -522,6 +541,30 @@ mod tests {
...
+            0.0 => 0.0,
+            1.0 => 0.0,
+            2.0 => 0.0,
+            3.0 => 0.0,
+            4.0 => 0.0,
+        }
+    }
+
+    test_beat! {
+        name: empty_arr,
```
Comment (porglezomp) [inline src/lib.rs]: Same for this, this is a static failure not a dynamic failure and so should produce an error on evaluation.

**Review by porglezomp — COMMENTED** [review]
Ah, I can't request changes because it's my own PR… hoisted by my own PR :)

File: src/lib.rs [hunk 5]
```diff
@@ -146,19 +146,22 @@ pub fn eval_beat(cmds: &[Cmd], t: f64) -> Result<f64, ()> {
...
+                // We need to pop size values, so our stack needs to be
+                // atleast size elements long. Note that split_off panics if we
+                // exceed the length of the vector, so we need this if guard.
+                if size > stack.len() { 
+                    return Err(())
+                } else if size == 0 {
+                    stack.push(0.0);
                 } else {
                     // We want to split off from the end, so we must subtract
                     // here. Note that we add one for the index.
```
Comment (porglezomp) [inline src/lib.rs]: This comment is out of date (we don't add one anymore)

File: src/lib.rs [hunk 6]
```diff
@@ -146,19 +146,22 @@ pub fn eval_beat(cmds: &[Cmd], t: f64) -> Result<f64, ()> {
...
-                    let split_index = stack.len() - (size + 1);
+                    let split_index = stack.len() - size;
                     let mut vec = stack.split_off(split_index);
                     let size = size as i32; // too lazy to keep tying as i32
                     // Pop instead of accessing it because the index shouldn't
```
Comment (porglezomp) [inline src/lib.rs]: This comment is also out of date

File: src/lib.rs [hunk 7]
```diff
@@ -146,19 +146,22 @@ pub fn eval_beat(cmds: &[Cmd], t: f64) -> Result<f64, ()> {
                 }
             }
             Arr(size) => {
-                // We need to pop index + size values, so our stack needs to be
-                // atleast size + 1 elements long
-                if size + 1 > stack.len() || size == 0 { 
-                    stack.push(0.0)
+                let index = stack.pop().ok_or(())? as i32;
```
Comment (porglezomp) [inline src/lib.rs]: Why use `i32` instead of `i64`?

~Alternately, does `as i64 as usize` give us the right behavior? `-1.0 as i64 as usize` gives `18446744073709551615` for me, which looks right. If this works nicely it would make some of the later casting unnecessary.~
Using `as i64 as usize` would give the wrong result for negative indices. `18446744073709551615 mod 3 = 0`, `-1 mod 3 = 2`.

File: src/lib.rs [hunk 8]
```diff
@@ -146,19 +146,22 @@ pub fn eval_beat(cmds: &[Cmd], t: f64) -> Result<f64, ()> {
...
+                    return Err(())
+                } else if size == 0 {
+                    stack.push(0.0);
                 } else {
                     // We want to split off from the end, so we must subtract
                     // here. Note that we add one for the index.
-                    let split_index = stack.len() - (size + 1);
+                    let split_index = stack.len() - size;
                     let mut vec = stack.split_off(split_index);
                     let size = size as i32; // too lazy to keep tying as i32
```
Comment (porglezomp) [inline src/lib.rs]: I don't think this `as` is worth commenting on, although I think it should go to `i64` not `i32`.

**Review by porglezomp — COMMENTED** [review]
Okay, some out-of-date comments, and casting.

**Review by porglezomp — COMMENTED** [review]
LGTM

**Review by a2aaron — APPROVED** [review]
